WIP: suppressing error boundary logs from intentionally thrown/caught errors #11098

vikramcse opened the pull request:
1. Removed support for the custom `suppressReactErrorLogging` property.
2. checkd e.defaultPrevented in our error event handler in `invokeGuardedCallback.js`.

vikramcse commented on packages/shared/invokeGuardedCallback.js:
```diff
@@ -173,6 +179,12 @@ if (__DEV__) {
+      if (shouldIgnoreError && error != null) {
+        this._hasSuppressedError = true;
+        this._suppressedError = error;
+        error.suppressReactErrorLogging = true;
```
while running yarn flow I am getting error `property `suppressReactErrorLogging`. Property not found in Error`.

vikramcse commented on packages/shared/invokeGuardedCallback.js:
```diff
@@ -22,6 +22,8 @@ let invokeGuardedCallback = function<A, B, C, D, E, F, Context>(
 ) {
   this._hasCaughtError = false;
   this._caughtError = null;
+  this._hasSuppressedError = false;
```
I did added `_hasSuppressedError` and `_suppressedError` to store the suppressed error and state, but can not figure out what to do with these fields.

gaearon commented on packages/shared/invokeGuardedCallback.js:
```diff
@@ -173,6 +179,12 @@ if (__DEV__) {
+      if (shouldIgnoreError && error != null) {
+        this._hasSuppressedError = true;
+        this._suppressedError = error;
+        error.suppressReactErrorLogging = true;
```
We should remove this concept completely.

gaearon replied:
This will require changing our unit test setup [here](https://github.com/facebook/react/blob/48833f698dc28a9af09819ba731a2e94d5bf9da7/scripts/jest/setupEnvironment.js#L43-L45) and [here](https://github.com/facebook/react/blob/48833f698dc28a9af09819ba731a2e94d5bf9da7/scripts/jest/setupEnvironment.js#L33-L37) not to rely on this property.

Instead we can try changing the `expect().toThrow()` Jest matcher to set up an `'error'` handler that does `preventDefault()`. Or figure out some other way to avoid failing all tests that intentionally throw an error.

vikramcse replied:
I added a functionality but I am not able to figure out the `resets instance variables before unmounting failed node` test is failing saying `Expected test not to call console.error().`
am i missing something?

gaearon commented on packages/react-reconciler/src/ReactFiberErrorLogger.js:
```diff
@@ -21,10 +21,6 @@ export function logCapturedError(capturedError: CapturedError): void {
   }
 
   const error = (capturedError.error: any);
-  const suppressLogging = error && error.suppressReactErrorLogging;
```
This should check `_hasSuppressedError`. I think in order to do that we'll want to remember `ReactErrorUtils. _hasSuppressedError` at the moment we capture the error (together with the `_caughtError` itself).

gaearon replied:
In other words we still want to *have a way* to suppress error logging.

We just don't want to use a custom property on the error object for that. Instead, we want to read a property on an object *we* created (`capturedError` is a custom object).

vikramcse replied:
I am able to get `_hasSuppressedError` property in `ReactFiberScheduler.js` and checked for the error, but in `ReactFiberErrorLogger.js` how do I get `_hasSuppressedError` property, as we have not imported `ReactErrorUtils.js`?

what I am thinking is to add a new property `hasSuppressedError` in  `capturedError` object 
```javascript
const capturedError = {
        componentName,
        componentStack,
        error,
        errorBoundary: errorBoundaryFound ? boundary.stateNode : null,
        errorBoundaryFound,
        errorBoundaryName,
        willRetry,
        hasSuppressedError,
      };
```
so that the can be found on `ReactFiberErrorLogger.js`.
is this a right approach?

gaearon replied:
Yeah that’s what I’m thinking.

vikramcse commented:
I added a functionality but I am not able to figure out the `resets instance variables before unmounting failed node` test is failing saying `Expected test not to call console.error().`
am i missing something?
```javascript
Expected test not to call console.error().
    
    If the warning is expected, test for it explicitly by:
    1. Using the .toWarnDev() / .toLowPriorityWarnDev() matchers, or...
    2. Mock it out using spyOnDev(console, 'error') or spyOnProd(console, 'error'), and test that the warning occurs.
    
    The above error occurred in the <Foo> component:
        in Foo (at ReactIncrementalErrorLogging-test.internal.js:229)
        in ErrorBoundary (at ReactIncrementalErrorLogging-test.internal.js:228)
    
    React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.
        at CustomConsole.newMethod [as error] (/home/vikram/code/javascript/oss/react/scripts/jest/setupTests.js:70:19)
        at logCapturedError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberErrorLogger.js:68:13)
        at captureError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1017:53)
        at renderRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:841:22)
        at performWorkOnRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1607:24)
        at performWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1517:7)
        at performAsyncWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1495:5)
        at flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:292:5)
        at flushUnitsOfWork.next (<anonymous>)
        at Object.flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:421:129)
      
      at Object.<anonymous> (scripts/jest/setupTests.js:114:15)

```

gaearon commented on packages/shared/ReactErrorUtils.js:
```diff
@@ -89,12 +92,21 @@ const ReactErrorUtils = {
       ReactErrorUtils._caughtError = null;
       ReactErrorUtils._hasCaughtError = false;
       return error;
+    } else if (ReactErrorUtils._hasSuppressedError) {
```
We should still count this as caught error. The only way `_hasSuppressedError` should be used is to remember this in `capturedError` so we can avoid logging it later. The rest of the logic should be the same.

vikramcse replied:
I did not understand!
In order to clear the `ReactErrorUtils._hasSuppressedError` we need to check the suppress error exist for not.
so count this as a `_caughtError` I need to set `ReactErrorUtils._hasCaughtError` true
```javascript
if (shouldIgnoreError && error != null) {
        this._hasSuppressedError = true;
        this._hasCaughtError = true;
        this._caughtError = error;
      }
```

gaearon replied:
Yes, that sounds good.

vikramcse replied:
also one of the test case if failing `resets instance variables before unmounting failed node` test is failing saying `Expected test not to call console.error().`
am i missing something?
```javascript
Expected test not to call console.error().
    
    If the warning is expected, test for it explicitly by:
    1. Using the .toWarnDev() / .toLowPriorityWarnDev() matchers, or...
    2. Mock it out using spyOnDev(console, 'error') or spyOnProd(console, 'error'), and test that the warning occurs.
    
    The above error occurred in the <Foo> component:
        in Foo (at ReactIncrementalErrorLogging-test.internal.js:229)
        in ErrorBoundary (at ReactIncrementalErrorLogging-test.internal.js:228)
    
    React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.
        at CustomConsole.newMethod [as error] (/home/vikram/code/javascript/oss/react/scripts/jest/setupTests.js:70:19)
        at logCapturedError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberErrorLogger.js:68:13)
        at captureError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1017:53)
        at renderRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:841:22)
        at performWorkOnRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1607:24)
        at performWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1517:7)
        at performAsyncWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1495:5)
        at flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:292:5)
        at flushUnitsOfWork.next (<anonymous>)
        at Object.flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:421:129)
      
      at Object.<anonymous> (scripts/jest/setupTests.js:114:15)

```

gaearon replied:
We want to change [this condition](https://github.com/facebook/react/blob/26185759e4f05025b1fdbe5e238a44aefe5669d4/scripts/jest/setupEnvironment.js#L43-L45) to always call `preventDefault()` (without the condition). Then it shouldn't log at all (since all errors will end up suppressed).

After that, a few other tests will fail that test *logging specifically*. We should figure out a way to exempt those.

vikramcse replied:
In out test cases we need to call `preventDefault` when any component throws error intentionally,
but I am not able to figure out how to add a error event handler in `Jest` as` window` object is not there.

gaearon replied:
I don't understand. Can you make a more detailed write-up about cases that are currently failing, and what you need help with?

vikramcse replied:
For example [test](https://github.com/facebook/react/blob/master/packages/react-test-renderer/src/__tests__/ReactTestRenderer-test.js#L398-L457) `supports error boundaries` in `ReactTestRenderer-test.js` failing. with an error `Expected test not to call console.error().` This is because we need to call `prevent default` so that out handleSuppressError logic gets executed. 

Here is an [example](https://jsfiddle.net/vikramcse/69z2wepo/97544/) and what I am thinking to add event handler.
But we can not do this because `window` object is not available.

gaearon replied:
>But we can not do this because as window object is not available in jsdom.

Not sure what you mean. It *is* available. There are a few tests that run without it because they run without jsdom in the first place, but generally it's there.

vikramcse replied:
In the past comments you mentioned (for writing unit test):

Instead we can try changing the `expect().toThrow()` Jest matcher to set up an 'error' handler that does `preventDefault()`. Or figure out some other way to avoid failing all tests that intentionally throw an error.

I am not able to figure out how to implement this, can you give me some clue?

gaearon replied:
You can add a custom matcher. See `toWarnDev` definition for an example of custom matcher. Maybe there is a way to similarly override the built-in `toThrow` matcher and do some custom logic in it?

vikramcse replied:
we have added below code in  `setupEnvironment.js` so that, when an error (intentionally) is thrown in our tests it will call `preventDefault` and eventually out `handleSuppress` logic will get called.

```javascript
if (typeof window !== 'undefined') {
  window.addEventListener('error', event => {
    if (event.error != null) {
      event.preventDefault();
    }
  });
```
But at the time of debugging in this `setupEnvironment.js` the `eventListnener` is not added because `window` object is `undefined`.

1. Is this  `setupEnvironment.js` file get used while running tests?
2. Why I am not able to get `window` object?

gaearon replied:
Some tests (only a few) use `@jest-environment node`. Search for it. Those won’t have a `window`, but they also won’t have this problem either (because the problem is specific to jsdom).

All other tests will have `window`.

Remember `setupEnvironment` executes once per test file, not once per project. So it runs in different environments.

vikramcse replied:
File `ReactTestRenderer.js` file has the `@jest-environment node` flag.  It means at the time of initialization `setupEnvironment` the `window` object is not available and therefore `eventListnener` is not initialized.

in which below test is failing.

```javascript
it('supports error boundaries', () => {
    const log = [];
    class Angry extends React.Component {
      render() {
        log.push('Angry render');
        throw new Error('Please, do not render me.');
      }
      componentDidMount() {
        log.push('Angry componentDidMount');
      }
      componentWillUnmount() {
        log.push('Angry componentWillUnmount');
      }
    }

    class Boundary extends React.Component {
      constructor(props) {
        super(props);
        this.state = {error: false};
      }
      render() {
        log.push('Boundary render');
        if (!this.state.error) {
          return (
            <div>
              <button onClick={this.onClick}>ClickMe</button>
              <Angry />
            </div>
          );
        } else {
          return <div>Happy Birthday!</div>;
        }
      }
      componentDidMount() {
        log.push('Boundary componentDidMount');
      }
      componentWillUnmount() {
        log.push('Boundary componentWillUnmount');
      }
      onClick() {
        /* do nothing */
      }
      componentDidCatch() {
        this.setState({error: true});
      }
    }

    const renderer = ReactTestRenderer.create(<Boundary />);
    expect(renderer.toJSON()).toEqual({
      type: 'div',
      props: {},
      children: ['Happy Birthday!'],
    });
    expect(log).toEqual([
      'Boundary render',
      'Angry render',
      'Boundary componentDidMount',
      'Boundary render',
    ]);
  });
```

error:
```javascript
Expected test not to call console.error().
    
    If the warning is expected, test for it explicitly by:
    1. Using the .toWarnDev() / .toLowPriorityWarnDev() matchers, or...
    2. Mock it out using spyOnDev(console, 'error') or spyOnProd(console, 'error'), and test that the warning occurs.
    
    The above error occurred in the <Foo> component:
        in Foo (at ReactIncrementalErrorLogging-test.internal.js:229)
        in ErrorBoundary (at ReactIncrementalErrorLogging-test.internal.js:228)
    
    React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.
        at CustomConsole.newMethod [as error] (/home/vikram/code/javascript/oss/react/scripts/jest/setupTests.js:70:19)
        at logCapturedError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberErrorLogger.js:68:13)
        at captureError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1017:53)
        at renderRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:841:22)
        at performWorkOnRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1607:24)
        at performWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1517:7)
        at performAsyncWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1495:5)
        at flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:292:5)
        at flushUnitsOfWork.next (<anonymous>)
        at Object.flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:421:129)
      
      at Object.<anonymous> (scripts/jest/setupTests.js:114:15)
```
Expected result would be error should not be printed in the console.
```javascript
function onError(event) {
        error = event.error;
        didSetError = true;
        if (error === null && event.colno === 0 && event.lineno === 0) {
          isCrossOriginError = true;
        }
        if (event.defaultPrevented) {
          shouldIgnoreError = true;
        }
      }
```
I checked the reason it's happening because in `invokeGuardedCallback.js`
the `onError` (above) function is not called so the `shouldIgnoreError` flag remains `false`.
Does this problem trace sounds good?

gaearon replied:
Yes, thanks!

I think for that test (and the few other tests that also have this problem) we can maybe wrap them in `expect(...).toWarnDev()`?

vikramcse replied:
By adding `expect(...).toWarnDev()` the tests are now passing  on my machine :)

vikramcse commented:
Hi, 
Here are the changes I made in my latest pull request

1. I have added `expect(...).toWarnDev()` [comment](https://github.com/facebook/react/pull/11981/commits/2cb99473216f5a590eee0ac4f184f6794c6b4087#r161211535) to my failing tests where test environment is `@node`. So at this point our `yarn test` is passing :)

2. But when I tried `yarn test-prod` most of the tests I fixed are failing.

vikramcse commented:
To solve the problem of tests failing in `prod` environment First I selected `ReactIncrementalErrorHandling-test.js` test to figure out the reason of failing the test.

the error is

```javascript
ReactIncrementalErrorHandling › catches render error in a boundary during full deferred mounting

    Expected test not to call console.error().
    
    If the warning is expected, test for it explicitly by:
    1. Using the .toWarnDev() / .toLowPriorityWarnDev() matchers, or...
    2. Mock it out using spyOnDev(console, 'error') or spyOnProd(console, 'error'), and test that the warning occurs.
    
    Error: Hello
        at Object.construct (/home/vikram/code/javascript/oss/react/scripts/jest/setupTests.js:149:29)
        at CustomConsole.newMethod [as error] (/home/vikram/code/javascript/oss/react/scripts/jest/setupTests.js:70:19)
        at logCapturedError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberErrorLogger.js:73:13)
        at captureError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1017:53)
        at renderRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:841:22)
        at performWorkOnRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1621:25)
        at performWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1517:7)
        at performAsyncWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1495:5)
        at flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:292:5)
        at flushUnitsOfWork.next (<anonymous>)
      
      at Object.construct (scripts/jest/setupTests.js:149:29)
      at Object.<anonymous> (scripts/jest/setupTests.js:114:15)
```
when I tried to put `spyOnProd(console, 'error')` it seems like it solved the problem.
so is it the right approach?

vikramcse commented:
Also I noticed in the tests files which `does not` have `@jest-environment node` failing in `yarn test-prod` but passes in `yarn test`.

In normal `yarn test` the error listener in `setupEnvironment.js` kicks in which calls `preventDefault` and our `suppress` logic works out as a result the error is suppressed. 

For example the file`multiple-copies-of-react-test.js` throwing error.
So am i also supposed to add `spyOnProd(console, 'error')` in these files? I am little confused? :)

vikramcse commented:
Hi @gaearon 
I am stuck in some tests which are failing in `yarn test-prod` mode. I have done some digging but I did not find any solution. Can I get little help?

vikramcse commented:
Hi, @gaearon
Is this issue(pull request) still relevent? I mean to say does someone else handled this issue his/her pull request. Actually I am stuck in writing some test cases [comment](https://github.com/facebook/react/pull/11981#issuecomment-360439637)

Thank You
